src/query/function_score_query.rs

```diff
@@ -0,0 +1,285 @@
+        FunctionScoreQuery { query, function }
+    }
+}
+
+impl Clone for FunctionScoreQuery {
+    fn clone(&self) -> Self {
+        FunctionScoreQuery {
+            query: self.query.box_clone(),
+            function: self.function.clone(),
+        }
+    }
+}
+
+impl fmt::Debug for FunctionScoreQuery {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "FunctionScore(query={:?})", self.query)
+    }
+}
+
+impl Query for FunctionScoreQuery {
+    fn weight(&self, enable_scoring: EnableScoring<'_>) -> crate::Result<Box<dyn Weight>> {
+        let inner_weight = self.query.weight(enable_scoring)?;
+        Ok(Box::new(FunctionScoreWeight::new(
+            inner_weight,
+            enable_scoring.is_scoring_enabled(),
+            self.function.clone(),
+        )))
+    }
+}
+
+struct FunctionScoreWeight {
+    weight: Box<dyn Weight>,
+    scoring_enabled: bool,
+    function: Function,
+}
+
+impl FunctionScoreWeight {
+    pub fn new(weight: Box<dyn Weight>, scoring_enabled: bool, function: Function) -> Self {
+        FunctionScoreWeight {
+            weight,
+            scoring_enabled,
+            function,
+        }
+    }
+}
+
+impl Weight for FunctionScoreWeight {
+    fn scorer(&self, reader: &SegmentReader, boost: Score) -> crate::Result<Box<dyn Scorer>> {
+        let inner_scorer = self.weight.scorer(reader, boost)?;
+        if !self.scoring_enabled {
+            Ok(Box::new(inner_scorer))
+        } else {
+            Ok(Box::new(FunctionScorer::new(
+                inner_scorer,
+                Arc::new(reader.clone()),
+                self.function.clone(),
+            )))
+        }
+    }
+
+    fn explain(&self, reader: &SegmentReader, doc: u32) -> crate::Result<Explanation> {
+        let mut scorer = self.scorer(reader, 1.0)?;
+        if scorer.doc() < doc && scorer.seek(doc) != doc {
+            return Err(TantivyError::InvalidArgument(format!(
+                "Document #({doc}) does not match"
+            )));
+        }
+        if !self.scoring_enabled {
+            return Ok(Explanation::new("FunctionScore with no scoring", 1.0));
+        }
+        let mut explanation = Explanation::new("FunctionScore", scorer.score());
+        let underlying_explanation = self.weight.explain(reader, doc)?;
+        explanation.add_detail(underlying_explanation);
+        Ok(explanation)
+    }
+
+    fn count(&self, reader: &SegmentReader) -> crate::Result<u32> {
+        self.weight.count(reader)
+    }
+}
+
+struct FunctionScorer {
+    scorer: Box<dyn Scorer>,
+    segment_reader: Arc<SegmentReader>,
+    function: Function,
+}
+
+impl FunctionScorer {
+    pub fn new(
+        scorer: Box<dyn Scorer>,
+        segment_reader: Arc<SegmentReader>,
+        function: Function,
+    ) -> Self {
+        FunctionScorer {
+            scorer,
+            segment_reader,
+            function,
+        }
+    }
+}
+
+impl DocSet for FunctionScorer {
```

Comment:
Can you implement the other methods too? seek for instance, has a default impemnetation but it will be catastrophically slow.